### PR TITLE
add `default-flavor` field to flutter pubspec, which will be used as the flavor in `flutter build/run` if `--flavor` is not provided

### LINKED ISSUE #22856: On Android when there are multiple flavors and no --flavor flag is specified then use a default one
On Android, when there are multiple flavors defined, and the user runs the app but does not specify a value for the `--flavor` flag then the following error is shown:

> You must specify a --flavor option to select one of them.

This is implemented like that here: https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/android/gradle.dart#L342

This makes it so that everytime we have to specifiy a value for `--flavor`.
I would propose when the flag is not specified to just default to the first one.
This will make it faster to start your Flutter app, not having to write `--flavor _flavor_` everytime.

### PR DESCRIPTION
**Pull request by holzgeist**
This PR adds a new flag `default-flavor` in the `flutter` section of `pubspec.yaml`. It allows developers of multi-flavor android apps to specify a default flavor to be used for `flutter run`, `flutter build` etc.
Using `flutter run` on flavored apps already works without specifying `--flavor` already works on iOS (it defaults to the `runner` schema), so I (and others in #22856) figured this would be nice to have.

fixes #22856

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[Data Driven Fixes]: https://github.com/flutter/flutter/wiki/Data-driven-Fixes

### CODE REVIEW

**Comment by holzgeist** [comment]
This probably needs to be reflected in https://github.com/flutter/website
(related to me not yet checking "I updated/added relevant documentation (doc comments with ///)." above

**Comment by holzgeist** [comment]
looks like the failing windows test OOMed:

```
[ +502 ms] > Task :app:packageDebug FAILED
[        ] AAPT2 aapt2-8.1.0-10154469-windows Daemon #0: shutdown
[        ] FAILURE: Build failed with an exception.
[        ] * What went wrong:
[        ] Execution failed for task ':app:packageDebug'.
[        ] > A failure occurred while executing com.android.build.gradle.tasks.PackageAndroidArtifact$IncrementalSplitterRunnable
[        ]    > java.lang.OutOfMemoryError (no error message)
```

This should not be related to my PR

**Comment by holzgeist** [comment]
ok, I realized now that this doesn't really work on iOS. The `runner` schema is only applied if there is no flavor. If there is one due to the default value, the build doesn't work anymore. Any ideas? Is there a way to tell if we are building for Android in `getBuildInfo` in `flutter_command`?

**Comment by holzgeist** [comment]
the solution to the ios problem is to rename the default build configuration from `Debug`, `Profile` and `Release` to `Debug <Default Flavor>`, etc.
Works like a charm :)

File: packages/flutter_tools/test/general.shard/runner/flutter_command_test.dart [hunk 1]
```diff
@@ -1217,6 +1217,46 @@ void main() {
...
+          return FlutterCommandResult.success();
+        },);
+        flutterCommand.usesFlavorOption();
+        final CommandRunner<void> runner =  createTestCommandRunner(flutterCommand);
+        await runner.run(<String>['dummy', '--flavor', 'bar']);
+        final BuildInfo buildInfo = await flutterCommand.getBuildInfo(forcedBuildMode: BuildMode.debug);
+        expect(buildInfo.flavor, 'bar');
+      }, overrides: <Type, Generator>{
+        FileSystem: () => fileSystem,
+        ProcessManager: () => FakeProcessManager.any(),
```
Comment (andrewkolos) [inline packages/flutter_tools/test/general.shard/runner/flutter_command_test.dart:1258]:
```suggestion
        ProcessManager: () => FakeProcessManager.empty(),
```

When possible, avoid `FakeProcessManager.any()` as it hides faked processes and can make tests more difficult to reason about.

File: packages/flutter_tools/test/general.shard/runner/flutter_command_test.dart [hunk 2]
```diff
@@ -1217,6 +1217,46 @@ void main() {
...
+flutter:
+  default-flavor: foo
+        ''');
+
+        final DummyFlutterCommand flutterCommand = DummyFlutterCommand();
+        final BuildInfo buildInfo = await flutterCommand.getBuildInfo(forcedBuildMode: BuildMode.debug);
+        expect(buildInfo.flavor, 'foo');
+      }, overrides: <Type, Generator>{
+        FileSystem: () => fileSystem,
+        ProcessManager: () => FakeProcessManager.any(),
```
Comment (andrewkolos) [inline packages/flutter_tools/test/general.shard/runner/flutter_command_test.dart:1235]:
```suggestion
        ProcessManager: () => FakeProcessManager.empty(),
```

**Review by andrewkolos — CHANGES_REQUESTED** [review]
Thanks for the PR! I think this feature will be simple enough to maintain and thus worth having in the tool. Everything in this PR looks correct. I just have a few nitpicks.

**Comment by andrewkolos** [comment]
I think we should document this new pubspec entry. We could do so on the flutter.dev page, but we could start by mentioning in the help text of `--flavor`:

https://github.com/flutter/flutter/blob/b5f6df7f1c0a6a1bbe60d420170a472ef0400baf/packages/flutter_tools/lib/src/runner/flutter_command.dart#L1093-L1100

What do you think about adding another line here that says something to the effect of

```
Overrides the value of the 'default-flavor' entry in the flutter pubspec.
```

**Comment by holzgeist** [comment]
Thanks @andrewkolos for the thorough review. I applied all your suggested changes.

Also, I added a (minor) mention to the `Flavors` documentation page: https://github.com/flutter/website/pull/10632

**Review by andrewkolos — APPROVED** [review]
LGTM!

**Comment by christopherfujino** [comment]
Looks like there are failing tests here

**Comment by christopherfujino** [comment]
> The test failures might be due to #148704.
> 
> Once the tree goes green again, I will update the PR branch to include [b6bed5a](https://github.com/flutter/flutter/commit/b6bed5aaf2c70e8c5d62a377797234beb923b3e6), which will skip the flaky tests.

oh nice

**Review by christopherfujino — APPROVED** [review]
RSLGTM

**Comment by andrewkolos** [comment]
> Also, I added a (minor) mention to the `Flavors` documentation page: [flutter/website#10632](https://github.com/flutter/website/pull/10632)

FYI https://github.com/flutter/website/pull/10632 got merged and then reverted since the functionality from this PR won't be available to stable release users until the next major release. Feel free to re-draft the PR and include note in the description that the feature won't be available until then (and thus the PR shouldn't be merged until then).

**Comment by holzgeist** [comment]
@andrewkolos oh, thanks. Sorry about that. I didn't think about that resp. I had no idea how fast documentation was released

I'll re-send the website PR

**Comment by navaronbracke** [comment]
@holzgeist Thanks for this! I did file https://github.com/flutter/flutter/issues/139286 in the past as well, which relates to the same problem from a tool perspective. If we get around to solving that issue - using a tool exit with a relevant message - perhaps we can include a reference to this fix in the message? 

For example: `The required flavor variant was not specified. Consider passing it as the "--flavor" option, or specifying the default flavor in the pubspec.`

**Comment by holzgeist** [comment]
> @holzgeist Thanks for this! I did file #139286 in the past as well, which relates to the same problem from a tool perspective. If we get around to solving that issue - using a tool exit with a relevant message - perhaps we can include a reference to this fix in the message?

Yup, an early exit would be better than the "no output file found" message in the end. Unfortunately I have no idea how to handle that

**Comment by andrewkolos** [comment]
> @andrewkolos it looks like my warning in the PR description wasn't enough, it got merged again: [flutter/website#10642](https://github.com/flutter/website/pull/10642)
> 
> Can you revert it again? I'll send a new PR once this feature hits stable

Oh no 🤭! Sorry about that. I've went ahead and gotten it reverted again (https://github.com/flutter/website/pull/10646). Per, https://github.com/flutter/website/pull/10646#issuecomment-2125660755, you are welcome to redraft it again now, but I would keep it marked as a [Draft](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request#converting-a-pull-request-to-a-draft) and include something like "[Next stable]" in the title.

> @holzgeist Thanks for this! I did file https://github.com/flutter/flutter/issues/139286 in the past as well, which relates to the same problem from a tool perspective. If we get around to solving that issue - using a tool exit with a relevant message - perhaps we can include a reference to this fix in the message?

Good idea. I've added a note about this idea on #139286 so that it doesn't get lost to history.

> Yup, an early exit would be better than the "no output file found" message in the end. Unfortunately I have no idea how to handle that

Fixing that issue would indeed be more complicated. I would love to do some digging and provide some helpful hints over on that thread, but alas I don't have the time at the moment 😞

Thanks again for the continued work here!

**Comment by sbatezat** [comment]
I don't get it.

Adding a "default-flavor" solves Android issue when it's not possible to "flutter build"/"flutter run" an app without flavor, but it introduces a new issue on iOS where, when uses, the iOS platform is now expecting the definition of a default-flavor (where it was not needed before).

Let's me explain further. Let's say i've got a "free" flavor and a normal app, like that :

> 
>     flavorDimensions "default"
> 
>     productFlavors {
>         FREE {
>             dimension "default"
>         }
>     }

I can fix the build of "flutter build" / "flutter run" without the flavor flag on the cli by adding your "default-flavor: PAID" flag on the pubspec with "PAID" also on the gradle file :

> 
>     flavorDimensions "default"
> 
>     productFlavors {
>         PAID {
>             dimension "default"
>         }
>         FREE {
>             dimension "default"
>         }
>     }

But then, on iOS, it's not working anymore (and it was working), because the "PAID" flavor does not exists. So I need to create it where it should not be needed....

Taking this into consideration, I'm not sure this MR is the way to go

**Comment by andrewkolos** [comment]
> I don't get it.

Unless https://github.com/flutter/flutter/issues/22856 was requesting the ability to configure a default flavor _exclusively_ for Android, this PR is a valid resolution to that issue. You are welcome to file a new feature request.

**Comment by antfitch** [comment]
Out of curiosity if a developer builds a Flutter app that ports to both iOS and Android, will this produce an error when the iOS app is built for that project?

**Comment by antfitch** [comment]
> > Out of curiosity, if a developer builds a Flutter app that ports to both iOS and Android, will this produce an error when the iOS app is built for that project?
> 
> Assuming "this" refers to the `default-flavor` feature/field in the pubspec, it should work for both platforms. However, it's very likely this feature doesn't work when building a Flutter app directly from Android Studio/Xcode (as opposed to using the `flutter` CLI or Flutter IDE plugin).

I agree, it makes sense that this wouldn't work from Android Studio/Xcode. I was noticing the thread of messages around iOS issues and the `default-flavor` field. I wanted to make sure that `default-flavor` in the pubspec works with flavors for iOS. It sounds like it does. Thanks!

**Pull request merged**